File: grpc/python/publisher.py
#!/usr/bin/env python3

import asyncio
import grpc
import pubsub_pb2
import pubsub_pb2_grpc
import sys

async def publish():
    if len(sys.argv) < 3:
        print(f"Usage: {sys.argv[0]} <topic> <values>")
        return
    topic = sys.argv[1]
    values_str = sys.argv[2]  # e.g., "1238,1235,5285,5456,5000,0,0"

    async with grpc.aio.insecure_channel('localhost:50051') as channel:
        stub = pubsub_pb2_grpc.PubSubServiceStub(channel)
        async def request_generator():
            # Send message
            msg = pubsub_pb2.Message(topic=topic)
            for val_str in values_str.split(','):
                val = pubsub_pb2.Value(int_value=int(val_str))
                msg.values.append(val)
            yield msg
            # Keep stream open
            while True:
                await asyncio.sleep(1)

        responses = stub.SubscribeAndPublish(request_generator())
        async for response in responses:
            print(f"Received on {response.topic}: {response.values}")

if __name__ == '__main__':
    asyncio.run(publish())

File: grpc/cpp/publisher.cpp
#include <grpcpp/grpcpp.h>
#include <grpcpp/channel.h>
#include <grpcpp/client_context.h>
#include <grpcpp/create_channel.h>
#include <grpcpp/security/credentials.h>
#include "pubsub.grpc.pb.h"
#include <iostream>
#include <thread>
#include <string>
#include <vector>

using grpc::Channel;
using grpc::ClientContext;
using grpc::ClientReaderWriter;
using grpc::Status;
using pubsub::PubSubService;
using pubsub::Message;
using pubsub::Value;

int main(int argc, char** argv) {
    if (argc < 3) {
        std::cerr << "Usage: " << argv[0] << " <topic> <message>" << std::endl;
        return 1;
    }
    std::string topic = argv[1];
    std::string msg_str = argv[2];

    auto channel = grpc::CreateChannel("localhost:50051", grpc::InsecureChannelCredentials());
    std::unique_ptr<PubSubService::Stub> stub = PubSubService::NewStub(channel);

    ClientContext context;
    std::shared_ptr<ClientReaderWriter<Message, Message>> stream(stub->SubscribeAndPublish(&context));

    // Thread to handle incoming messages (optional for publisher)
    std::thread reader_thread([stream]() {
        Message response;
        while (stream->Read(&response)) {
            std::cout << "Received on " << response.topic() << ": ";
            for (const auto& value : response.values()) {
                if (value.has_string_value()) {
                    std::cout << value.string_value() << " ";
                } else if (value.has_int_value()) {
                    std::cout << value.int_value() << " ";
                } else if (value.has_double_value()) {
                    std::cout << value.double_value() << " ";
                }
            }
            std::cout << std::endl;
        }
    });

    // Send message
    Message msg;
    msg.set_topic(topic);
    Value* val = msg.add_values();
    val->set_string_value(msg_str);
    stream->Write(msg);

    // Keep stream open for a bit
    std::this_thread::sleep_for(std::chrono::seconds(1));

    stream->WritesDone();
    Status status = stream->Finish();
    if (!status.ok()) {
        std::cerr << "RPC failed: " << status.error_message() << std::endl;
        return 1;
    }

    reader_thread.join();
    return 0;
}
